src/passkey_child/passkey_child_assert.c

```diff
@@ -0,0 +1,271 @@
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+errno_t
+set_assert_options(fido_opt_t up, fido_opt_t uv, fido_assert_t *_assert)
+{
+    errno_t ret;
+
+    ret = fido_assert_set_up(_assert, up);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+            "fido_assert_set_up failed [%d]: %s.\n",
+            ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = fido_assert_set_uv(_assert, uv);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_assert_set_uv failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+errno_t
+prepare_assert(const struct passkey_data *data, int index,
+               fido_assert_t *_assert)
+{
+    TALLOC_CTX *tmp_ctx = NULL;
+    unsigned char *key_handle;
+    size_t key_handle_len;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = fido_assert_set_rp(_assert, data->domain);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_assert_set_rp failed [%d]: %s.\n", ret, fido_strerr(ret));
+        goto done;
+    }
+
+    key_handle = sss_base64_decode(tmp_ctx, data->key_handle_list[index],
+                                   &key_handle_len);
+    if (key_handle == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to decode key handle.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = fido_assert_allow_cred(_assert, key_handle, key_handle_len);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_assert_allow_cred failed [%d]: %s.\n", ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = set_assert_options(FIDO_OPT_FALSE, FIDO_OPT_OMIT, _assert);
+    if (ret != FIDO_OK) {
+        goto done;
+    }
+
+    ret = set_assert_client_data_hash(_assert);
+    if (ret != EOK) {
+        goto done;
+    }
+
+done:
+    talloc_free(tmp_ctx);
+
+    return ret;
+}
+
+errno_t
+reset_public_key(struct passkey_data *_data)
+{
+    if (_data->type == COSE_ES256) {
+        es256_pk_free((es256_pk_t **) &_data->public_key);
+    } else if (_data->type == COSE_RS256) {
+        rs256_pk_free((rs256_pk_t **) &_data->public_key);
+    } else if (_data->type == COSE_EDDSA) {
+        eddsa_pk_free((eddsa_pk_t **) &_data->public_key);
+    }
+
+    return EOK;
+}
+
+errno_t
+decode_public_key(char *b64_public_key, struct passkey_data *_data)
+{
+    TALLOC_CTX *tmp_ctx = NULL;
+    unsigned char *buf = NULL;
+    size_t buf_len;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    buf = sss_base64_decode(tmp_ctx, b64_public_key, &buf_len);
+    if (buf == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to decode public key.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    if (_data->type == COSE_ES256) {
```

Comment:
It looks like the type is expected to be the same over all public keys. What if I have one EDDSA key and one ES256 key?

Reply:
I changed what we will print when registering the key and what will be stored in the LDAP server. Nowadays, the public key will be printed and stored in PEM format. This way we can obtain the COSE type from OpenSSL and proceed accordingly.